File: backend/app/schemas/pulse_source.py
from datetime import datetime
from typing import Optional

from pydantic import BaseModel


class PulseSourceCreate(BaseModel):
    telegram_id: int
    username: Optional[str] = None
    title: str
    category: str
    subcategory: Optional[str] = None
    keywords: Optional[list[str]] = None
    criteria: Optional[dict] = None


class PulseSourceUpdate(BaseModel):
    title: Optional[str] = None
    category: Optional[str] = None
    subcategory: Optional[str] = None
    keywords: Optional[list[str]] = None
    criteria: Optional[dict] = None
    is_active: Optional[bool] = None


class PulseSourceResponse(BaseModel):
    id: int
    user_id: int
    telegram_id: int
    username: Optional[str] = None
    title: str
    category: str
    subcategory: Optional[str] = None
    keywords: Optional[list[str]] = None
    criteria: Optional[dict] = None
    is_active: bool
    last_polled_at: Optional[datetime] = None
    poll_status: str = "idle"
    last_poll_error: Optional[str] = None
    last_poll_message_count: int = 0
    created_at: datetime

    model_config = {"from_attributes": True}


class PollStatusSourceResponse(BaseModel):
    id: int
    title: str
    poll_status: str
    last_poll_error: Optional[str] = None
    last_poll_message_count: int = 0
    last_polled_at: Optional[datetime] = None

    model_config = {"from_attributes": True}


class PollStatusResponse(BaseModel):
    sources: list[PollStatusSourceResponse]
    any_polling: bool


class PulseSourceResolveResponse(BaseModel):
    telegram_id: int
    username: Optional[str] = None
    title: str
    members_count: Optional[int] = None

File: frontend/src/types/pulse-source.ts
export interface PulseSource {
  id: number;
  user_id: number;
  telegram_id: number;
  username: string | null;
  title: string;
  category: string;
  subcategory: string | null;
  keywords: string[] | null;
  criteria: Record<string, unknown> | null;
  is_active: boolean;
  last_polled_at: string | null;
  poll_status: "idle" | "polling" | "error";
  last_poll_error: string | null;
  last_poll_message_count: number;
  created_at: string;
}

export interface PollStatusSource {
  id: number;
  title: string;
  poll_status: "idle" | "polling" | "error";
  last_poll_error: string | null;
  last_poll_message_count: number;
  last_polled_at: string | null;
}

export interface PollStatusResponse {
  sources: PollStatusSource[];
  any_polling: boolean;
}

export interface PulseSourceCreate {
  telegram_id: number;
  username?: string;
  title: string;
  category: string;
  subcategory?: string;
  keywords?: string[];
  criteria?: Record<string, unknown>;
}

export interface PulseSourceUpdate {
  title?: string;
  category?: string;
  subcategory?: string;
  keywords?: string[];
  criteria?: Record<string, unknown>;
  is_active?: boolean;
}

export interface PulseSourceResolveResult {
  telegram_id: number;
  username: string | null;
  title: string;
  members_count: number | null;
}

export const SOURCE_CATEGORIES = ["news", "jobs", "learning"] as const;

export const CATEGORY_LABELS: Record<string, string> = {
  news: "News",
  jobs: "Jobs",
  learning: "Learning",
};
